### PR TITLE
[WIP] Add initial support for updating systemd slices

### LINKED ISSUE #88197: kubelet incorrectly configures `kubepods.slice` systemd unit when kube-reserved and/or system-reserved is used
**What happened**:

Kubelet is started on CentOS 7 with systemd with the following switches (among others) `--cgroup-driver=systemd`, `--enforce-node-allocatable=pods`, `--kube-reserved=cpu=100m,ephemeral-storage=2Gi,memory=1.5Gi`

With these settings kubelet correctly creates `kubepods.slice` cgroup with memory limit set to total RAM on the node (~4Gi in this example) minus kube-reserved (1.5Gi in this example)
```
# free -b
              total        used        free      shared  buff/cache   available
Mem:     3973447680   167432192  3489378304      528384   316637184  3584466944

# cat /sys/fs/cgroup/memory/kubepods.slice/memory.limit_in_bytes
2362834944
```

At the same time corresponding systemd unit `kubepods.slice` created by kubelet has a `MemoryLimit` property incorrectly set to total RAM of the node (without subtracting kube-reserved amount):
```
# systemctl show kubepods.slice
Slice=-.slice
ControlGroup=/kubepods.slice
...
MemoryLimit=3973447680
...

# systemctl cat kubepods.slice
...
# /run/systemd/system/kubepods.slice.d/50-MemoryLimit.conf
[Slice]
MemoryLimit=3973447680
...
```

This leads to systemd sometimes resetting the limit in `/sys/fs/cgroup/memory/kubepods.slice/memory.limit_in_bytes` cgroup file to total RAM on the node without accounting for the kube-reserved (and/or system-reserved) amount.
This in turn can lead to pods using more memory than expected and OOM kill events, and unstable node behavior in general

**What you expected to happen**:

`kubepods.slice` systemd unit `MemoryLimit` property should be set to the same value as `/sys/fs/cgroup/memory/kubepods.slice/memory.limit_in_bytes` taking into account kube-reserved and system-reserved amounts.

When kubelet s restarted with different kube-reserved and/or system-reserved values, both `/sys/fs/cgroup/memory/kubepods.slice/memory.limit_in_bytes` and `MemoryLimit` property of `kubepods.slice` systemd unit should be changed to the same value even if the cgroups and the systemd unit already exist.

**How to reproduce it (as minimally and precisely as possible)**:

Run kubelet on any systemd linux with the switches specified above and observe the discrepancy between memory limits defined in `/sys/fs/cgroup/memory/kubepods.slice/memory.limit_in_bytes` and `MemoryLimit` property of `kubepods.slice` systemd unit.

**Anything else we need to know?**:

**Environment**:
- Kubernetes version (use `kubectl version`):
  1.16.0, but also applies to all earlier and later versions
- Cloud provider or hardware configuration:
   virtualbox VM, but applies to any infrastructure
- OS (e.g: `cat /etc/os-release`):
   centos 7, but applies to any systemd based linux flavor
- Kernel (e.g. `uname -a`):
   `3.10.0-1062.9.1.el7.x86_64 #1 SMP Fri Dec 6 15:49:49 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux` but applies to any other kernel
- Install tools:
   kublr, but applies to any other tool
- Network plugin and version (if this is a network-related bug): N/A
- Others: N/A

### PR DESCRIPTION
**Pull request by odinuge**
**What type of PR is this?**

/kind bug

**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:

Fixes #88197

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
Add support for updating resources parameters for the different qos classes when using cgroup driver systemd
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs

```

### CODE REVIEW

**Comment by rphillips** [comment]
cc @sjenning

**Comment by rphillips** [comment]
I think the update should be in the opencontainers/runc Manager. Is that not possible?

**Comment by odinuge** [comment]
> @rphillips I think the update should be in the opencontainers/runc Manager. Is that not possible?

I do agree that this code may fit better outside the k/k repo, but I am not sure if it would fit in runc/libcontainer. The way we use `runc/libcontainer` (in this case) for controling cgroups and systemd slices is outside the scope of runc as I understand it, so adding it there would require some time and extra work for the initial implementation. From my experience getting things like this into runc is hard and cumbersome, and we already do quite a lot of the work ourselves inside `cgroup_manager_linux`. Eg. we have to update the cgroup subsystems manually ref. comment https://github.com/kubernetes/kubernetes/pull/88404/files#diff-fa76bb6ae2d9b4bb3d023737fe5e6029L347. 

Some of the code in `cgroup_manager_linux` can now be migrated into code using `run/libcontainer`, but I think the best approach would be to fix it in k/k and write some proper e2e tests for it, making it easier to switch to using vendored code, and adding the functionality to `runc/libcontainer`. There is currently no e2e tests running with `--cgroup-driver=systemd` (unsure about openshift tho), so imo. that would be the first place to start.

But that is just my thoughts tho.

**Comment by odinuge** [comment]
Ref. cgroupv2 implemenetation discussions:
/cc AkihiroSuda giuseppe

**Comment by AkihiroSuda** [comment]
The amount of the code copied from runc is relatively huge, so I prefer this PR to be applied to runc repo as well, but not a blocker, I think.

File: pkg/kubelet/cm/cgroup_manager_linux.go [hunk 1]
```diff
@@ -440,12 +453,81 @@ func (m *cgroupManagerImpl) Update(cgroupConfig *CgroupConfig) error {
...
+		unitName   = c.Name
+		properties []systemddbus.Property
+	)
+
+	if c.Resources.Memory != 0 {
+		properties = append(properties,
+			newProp("MemoryLimit", uint64(c.Resources.Memory)))
+	}
+
+	if c.Resources.CpuShares != 0 {
```
Comment (giuseppe) [inline pkg/kubelet/cm/cgroup_manager_linux.go]: this won't work on cgroup v2.

We would need to set the `CPUWeight=` property (https://github.com/kubernetes/kubernetes/pull/89567 moves the conversion to a new function so should be easier to use)

File: pkg/kubelet/cm/cgroup_manager_linux.go [hunk 2]
```diff
@@ -440,12 +453,81 @@ func (m *cgroupManagerImpl) Update(cgroupConfig *CgroupConfig) error {
...
+			cpuQuotaPerSecUSec = uint64(c.Resources.CpuQuota*1000000) / c.Resources.CpuPeriod
+			if cpuQuotaPerSecUSec%10000 != 0 {
+				cpuQuotaPerSecUSec = ((cpuQuotaPerSecUSec / 10000) + 1) * 10000
+			}
+		}
+		properties = append(properties,
+			newProp("CPUQuotaPerSecUSec", cpuQuotaPerSecUSec))
+	}
+
+	if c.Resources.BlkioWeight != 0 {
```
Comment (giuseppe) [inline pkg/kubelet/cm/cgroup_manager_linux.go]: is it currently used by the kubelet or can we drop it?  For cgroup v2 we would need to convert to a different range and use the `IOWeight=` property

**Comment by odinuge** [comment]
Fixed in runc, closing in favor of #90489

**Pull request closed without merging**